transition.go

```diff
@@ -0,0 +1,142 @@
+}
+
+const (
+	// no transition is happening
+	ToNone Transition = iota
+	// initial transition
+	ToIdle
+	ToSync
+	ToError
+	ToDownload
+	ToArtifactInstall
+	// should hsve Enter and Error actions
+	ToArtifactReboot_Enter
+	// should have Leave action only
+	ToArtifactReboot_Leave
+	ToArtifactCommit
+	ToArtifactRollback
+	// should hsve Enter and Error actions
+	ToArtifactRollbackReboot_Enter
+	// should have Leave action only
+	ToArtifactRollbackReboot_Leave
+	ToArtifactError
+)
+
+var (
+	transitionNames = map[Transition]string{
+		ToNone:                         "none",
+		ToIdle:                         "Idle",
+		ToSync:                         "Sync",
+		ToError:                        "Error",
+		ToDownload:                     "Download",
+		ToArtifactInstall:              "ArtifactInstall",
+		ToArtifactReboot_Enter:         "ArtifactReboot_Enter",
+		ToArtifactReboot_Leave:         "ArtifactReboot_Leave",
+		ToArtifactCommit:               "ArtifactCommit",
+		ToArtifactRollback:             "ArtifactRollback",
+		ToArtifactRollbackReboot_Enter: "ArtifactRollbackReboot_Enter",
+		ToArtifactRollbackReboot_Leave: "ArtifactRollbackReboot_Leave",
+		ToArtifactError:                "ArtifactError",
+	}
+)
+
+func (t Transition) String() string {
+	return transitionNames[t]
+}
+
+// For some states we should ignore errors as recovery is not possible
+// and we might end up with device bricked.
+func ignoreErrors(t Transition, action string) bool {
+	return t == ToIdle ||
+		t == ToArtifactRollback ||
+		t == ToArtifactRollbackReboot_Enter ||
+		t == ToArtifactError ||
+		action == "Error"
+}
+
+// Transition implements statescript.Launcher interface
+func (t Transition) Enter(exec statescript.Executor) error {
+	if t == ToNone {
+		return nil
+	}
+
+	name := t.String()
+	spl := strings.Split(name, "_")
+	if len(spl) == 2 {
+		name = spl[0]
+		if spl[1] != "Enter" {
+			return nil
+		}
+	}
+
+	if err := exec.ExecuteAll(name, "Enter", ignoreErrors(t, "Enter")); err != nil {
+		return errors.Wrapf(err, "error running enter state script(s) for %v state", t)
+	}
+	return nil
+}
+
+func (t Transition) Leave(exec statescript.Executor) error {
+	if t == ToNone {
+		return nil
+	}
+
+	name := t.String()
+	spl := strings.Split(name, "_")
+	if len(spl) == 2 {
+		name = spl[0]
+		if spl[1] != "Leave" {
+			return nil
+		}
+	}
+
+	if err := exec.ExecuteAll(name, "Leave", ignoreErrors(t, "Leave")); err != nil {
+		return errors.Wrapf(err, "error running leave state script(s) for %v state", t)
+	}
+	return nil
+}
+
+func (t Transition) Error(exec statescript.Executor) error {
+	if t == ToNone {
+		return nil
+	}
+
+	name := t.String()
+	spl := strings.Split(name, "_")
+	if len(spl) == 2 {
+		name = spl[0]
+		if spl[1] != "Leave" {
+			return nil
+		}
+	}
+
+	if err := exec.ExecuteAll(name, "Error", ignoreErrors(t, "Error")); err != nil {
+		return errors.Wrapf(err, "error running error state script(s) for %v state", t)
+	}
+	return nil
+}
```

Comment:
Is the only difference between these three functions the words "Enter", "Leave" and "Error"? I think you could consolidate these by having an extra argument with this word in it. Possibly with three wrappers to call it. Less code means less maintenance!